### PR TITLE
add path prefix to ObjectStorage.Iterator

### PR DESCRIPTION
**Pull request by fuxiaohei**
Support to iterator subdirectory in ObjectStorage for ObjectStorage.Iterator method.

It's required for https://github.com/go-gitea/gitea/pull/22738 to make artifact files cleanable.

### CODE REVIEW

File: modules/storage/local.go [hunk 1]
```diff
@@ -128,8 +128,13 @@ func (l *LocalStorage) URL(path, name string) (*url.URL, error) {
 }
 
 // IterateObjects iterates across the objects in the local storage
-func (l *LocalStorage) IterateObjects(fn func(path string, obj Object) error) error {
-	return filepath.WalkDir(l.dir, func(path string, d os.DirEntry, err error) error {
+func (l *LocalStorage) IterateObjects(prefix string, fn func(path string, obj Object) error) error {
+	dir := l.dir
+	prefix = strings.TrimSpace(prefix)
+	if prefix != "" {
+		dir = filepath.Join(strings.TrimSuffix(l.dir, "/"), "/"+strings.TrimPrefix(prefix, "/"))
```
Comment (wolfogre) [inline modules/storage/local.go]: Maybe
```suggestion
		dir = filepath.Join(l.dir, prefix))
```
is enough.

And what if `IterateObjects("../../../etc/passwd")`?

File: modules/storage/local.go [hunk 2]
```diff
@@ -128,8 +128,13 @@ func (l *LocalStorage) URL(path, name string) (*url.URL, error) {
 }
 
 // IterateObjects iterates across the objects in the local storage
-func (l *LocalStorage) IterateObjects(fn func(path string, obj Object) error) error {
-	return filepath.WalkDir(l.dir, func(path string, d os.DirEntry, err error) error {
+func (l *LocalStorage) IterateObjects(prefix string, fn func(path string, obj Object) error) error {
+	dir := l.dir
+	prefix = strings.TrimSpace(prefix)
```
Comment (wolfogre) [inline modules/storage/local.go]: It's a friendly correction, however, what if I really have a file with path `/a/b /c.txt`? It's fine for both filesystem and object storage.

Comment (fuxiaohei) [reply modules/storage/local.go]: Yes. blank space is correct in filesystem. no need to trim.

File: modules/storage/local.go (hunk 1 above)
Comment (fuxiaohei) [reply modules/storage/local.go]:
```
if !strings.HasPrefix(dir, strings.TrimSuffix(l.dir, "/")) {
	return ErrIllegalPath
}
```

I check the joined directory. It must be prefixed with storage directory.

File: modules/storage/local.go [hunk 3]
```diff
@@ -128,8 +128,16 @@ func (l *LocalStorage) URL(path, name string) (*url.URL, error) {
...
 
 // IterateObjects iterates across the objects in the local storage
-func (l *LocalStorage) IterateObjects(fn func(path string, obj Object) error) error {
-	return filepath.WalkDir(l.dir, func(path string, d os.DirEntry, err error) error {
+func (l *LocalStorage) IterateObjects(prefix string, fn func(path string, obj Object) error) error {
+	dir := l.dir
+	if prefix != "" {
+		prefix_trim := filepath.Clean(prefix)
+		dir = filepath.Join(l.dir, prefix_trim)
+		if !strings.HasPrefix(dir, strings.TrimSuffix(l.dir, "/")) {
```
Comment (lunny) [inline modules/storage/local.go]: Looks like it's useless except you use `filepath.Clean` on the dir

Comment (fuxiaohei) [reply modules/storage/local.go]: Yeah. filepath.Join seems shorten result path as Clean method called.

File: modules/storage/local.go [hunk 4]
```diff
@@ -128,8 +128,15 @@ func (l *LocalStorage) URL(path, name string) (*url.URL, error) {
 }
 
 // IterateObjects iterates across the objects in the local storage
-func (l *LocalStorage) IterateObjects(fn func(path string, obj Object) error) error {
-	return filepath.WalkDir(l.dir, func(path string, d os.DirEntry, err error) error {
+func (l *LocalStorage) IterateObjects(prefix string, fn func(path string, obj Object) error) error {
+	dir := l.dir
+	if prefix != "" {
+		dir = filepath.Join(l.dir, prefix)
```
Comment (lunny) [inline modules/storage/local.go]:
```suggestion
		dir = filepath.Clean(filepath.Join(l.dir, prefix))
```

**Comment by wolfogre** [comment]
I think #23371 can help.

File: modules/storage/local.go (hunk 4 above)
Comment (fuxiaohei) [reply modules/storage/local.go]: use util.CleanPath from #23371

**Review by lunny — APPROVED** [review]

**Review by wolfogre — APPROVED** [review]

**Pull request merged**